libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -517,6 +517,108 @@ string YulUtilFunctions::arrayLengthFunction(ArrayType const& _type)
+	solUnimplementedAssert(!_type.isByteArray(), "Byte Arrays not yet implemented!");
+
+	if (!_type.isByteArray() && _type.baseType()->storageBytes() < 32)
+		solAssert(_type.baseType()->isValueType(), "Invalid storage size for non-value type.");
+
+	string functionName = "resize_array_" + _type.identifier();
+	return m_functionCollector->createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(array, newLen) {
+				let oldLen := <fetchLength>(array)
+
+				// Store new length
+				sstore(array, newLen)
+
+				// Size was reduced, clear end of array
+				if lt(newLen, oldLen) {
+					let oldSize := <convertToSize>(oldLen)
+					let newSize := <convertToSize>(newLen)
+					let arrayDataStart := <dataPosition>(array)
+					let deleteStart := add(arrayDataStart, newSize)
+					let deleteEnd := add(arrayDataStart, oldSize)
+					<clearStorageRange>(deleteStart, deleteEnd)
+				}
+			})")
+			("functionName", functionName)
+			("fetchLength", arrayLengthFunction(_type))
+			("convertToSize", arrayConvertLengthToSize(_type))
+			("dataPosition", arrayDataAreaFunction(_type))
+			("clearStorageRange", clearStorageRangeFunction(*_type.baseType()))
+			.render();
+	});
+}
+
+string YulUtilFunctions::clearStorageRangeFunction(Type const& _type)
+{
+	string functionName = "clear_storage_range_" + _type.identifier();
+
+	solUnimplementedAssert(_type.storageSize() == 1, "");
+
+	return m_functionCollector->createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(start, end) {
+				for {} iszero(gt(start, end)) { start := add(start, <storageSize>) }
+				{
+					sstore(start, <zeroValue>())
+				}
+			}
+		)")
+		("functionName", functionName)
+		("updateStorageValue", updateStorageValueFunction(_type))
+		("zeroValue", zeroValueFunction(_type))
+		("storageSize", _type.storageSize().str())
+		.render();
+	});
+}
+
+string YulUtilFunctions::arrayConvertLengthToSize(ArrayType const& _type)
+{
+	string functionName = "array_convert_length_to_size_" + _type.identifier();
+	return m_functionCollector->createFunction(functionName, [&]() {
+		Type const& baseType = *_type.baseType();
+		unsigned const baseBytes = baseType.storageBytes();
+
+		return Whiskers(R"(
+			function <functionName>(length) -> size {
+				size := length
+				<?storage>
+					<?storageSizeLessEq1>
+						<?baseBytesIs0>
+							size := 1
+						</baseBytesIs0>
+						<?baseBytesLessEq16>
+							size := div(<itemsPerSlot>, <add>(length, <sub>(<itemsPerSlot>, 1)))
+						</baseBytesLessEq16>
+					<!storageSizeLessEq1>
+						size := <mul>(<storageSize>, length)
+					</storageSizeLessEq1>
+				<!storage>
+					<?notByteArray>
+						size := <mul>(length, <elementSize>)
+					</notByteArray>
+				</storage>
+			})")
+			("functionName", functionName)
+			("storage", _type.location() == DataLocation::Storage)
+			("storageSizeLessEq1", baseType.storageSize() <= 1)
+			("baseBytesIs0", baseBytes == 0)
```

Comment:
Can this happen?

Reply:
This is more or less a direct translation from https://github.com/ethereum/solidity/blob/develop/libsolidity/codegen/ArrayUtils.cpp#L972
which does include exactly such a check.

Reply:
I think nowadays, we do not have such types anymore. Please add an assertion.